src/V3AstNodeExpr.h

```diff
@@ -451,6 +451,7 @@ class AstNodeVarRef VL_NOT_FINAL : public AstNodeExpr {
     AstNodeModule* m_classOrPackagep = nullptr;  // Package hierarchy
     string m_name;  // Name of variable
     string m_selfPointer;  // Output code object pointer (e.g.: 'this')
+    bool m_underGenerate = false;
```

Comment:
We should avoid adding any "under's" in VarRefs because they are often cloned and move around, so may end up incorrect (not under when they are, or vice-versa).

See comments below too.

Reply:
Are you suggesting to rename, or to move to a different Ast class besides `AstNodeVarRef`? Putting it here seemed to behave nicely when I ran the regression suite. I experimented with a handful of different ways to identify in V3Undriven alone whether a given `AstNodeVarRef` was within a `generate`, and nothing seemed to work. E.g. I tried creating `visit(AstGenFor*)` and `visit(AstGenIf*)` and `visit (AstBegin*)` functions with V3Undriven, but the behavior didn't make sense--it seemed that the state of being 'underGenerate' didn't persist as the nodes were iterated. It was only when I started using V3LinkResolve instead that the state was persistent. V3LinkResolve already had a `visit(AstNodeVarRef*)` function and already had a `m_underGenerate` member, and of course my check is for `AstNodeVarRef`s, so it all worked quite nicely.

Reply:
I'd prefer this not to be tracked by the AstTree at all.   See other comment for another possibility using fileline.

Reply:
Using fileline I was able to prevent firing on most cases under a generate. But there's one (legal) case I'm still not yet preventing firing on:
```
generate
  always_comb out[0] = d;
  always_comb out[1] = d;
endgenerate
```

It's difficult to distinguish it from this (illegal):
```
always_comb out = d;
always_comb out = d;
```

The latter I'm able to catch by noticing they're the same wholly-driven vars but in different always_comb processes. However, the former is also 'wholly-driven' vars in different always_comb processes. As I mentioned in a previous comment, V3Undriven is treating `out[0]=d`  as 'wholly driven' if under a generate.

Any suggestions would be greatly appreciated. Otherwise I'll keep tinkering

Reply:
V3Undriven does bit-by-bit checking for normal UNDRIVEN, perhaps the extensions you are adding aren't also doing bit-by-bit?

Reply:
So I'm intentionally not doing bit-by-bit because per 9.2.2.2 in the spec,
```
always_comb out[0] = d0;
always_comb out[1] = d1;
```
is legal, and V3Undriven correctly doesn't consider `out[0]=d0` as making `out` 'wholly driven'. Hence my check doesn't misfire on the above. However, if I wrap the above code in a `generate`, V3Undriven suddenly starts treating `out[0]=d0` as making `out` 'wholly driven', and my check misfires.
I might be misunderstanding something, but there seems to be a fundamental difference in how V3Undriven is treating vars under a `generate` vs. those same vars if they weren't under a `generate`

Reply:
There's an existing check in V3Undriven for `ALWCOMBORDER`, and that check seems to avoid this problem by simply turning off linting for generates in the tests.  `t_split_var_0.v` and `t_unpack_array_no_expand.v` have examples of this

Reply:
When you said varrefs are often cloned or move around, and hence not a good place to put an 'under' member, does that cloning/moving happen _after_ V3Undriven? If so, possibly I could go back to  having an `underGenerate`  signal in the `AstNodeVarRef` ( which was my original solution that passed all the tests), but with the following improvements:

1. change the name to something else such as `preUndrivenUnderGenerate` 
2. make it completely private, except `UndrivenVisitor` could be a c++ `friend` who could access it

Just throwing out an idea in case we hit a roadblock with the other solutions

Reply:
The only disables for generates I see is "!isGenVar", this is saying it's a "genvar", it is not checking a variable it is inside a generate.

I really don't want anything added to VarRef.

Which case is giving you problems? The AST is as I'd expect and similar under generates or not for this:

```
module x;
   reg [1:0] o,og,u,ug;
   always_comb o[0] = 1;
   always_comb o[1] = 1;
   always_comb u[0] = 1;
   always_comb u[1] = 1;
   generate begin
      always_comb og[0] = 1;
      always_comb og[1] = 1;
      always_comb ug = 1;
      always_comb ug = 1;
   end
   endgenerate
en
```

Reply:
> The only disables for generates I see is "!isGenVar", this is saying it's a "genvar", it is not checking a variable it is inside a generate.
Yep, I understood this. I wasn't talking about the check itself. I was saying there are `verilator lint_off ALWCOMBORDER` directives in `t_split_var_0.v` and `t_unpack_array_no_expand.v` which allow those tests to pass despite  `ALWCOMBORDER` not tracking if under a generate.  See here: https://github.com/verilator/verilator/blob/master/test_regress/t/t_unpack_array_no_expand.v#L13
This is also the only remaining case that is giving me trouble with my check. `stage_data` in this `t_unpack_array_no_expand.v`

Reply:
Apologies, for this last case I realized the `generate` isn't what's causing the problem now, it's the unpacked array. I was able to narrow it down to a simpler example:
```
reg d_unpacked [1:0];
always_comb d_unpacked[0] = 1; 
always_comb d_unpacked[1] = 1; //<---legal, but my check incorrectly misfires

reg [1:0] d_packed;
always_comb d_packed[0] = 1;
always_comb d_packed[1] = 1; //<---legal, my check correctly doesn't fire

```
I think I just need to make my check a bit smarter to identify unpacked arrays

Reply:
I just updated pull request with a solution that doesn't involve any tracking in the AstTree, and is self-contained within V3Undriven. I did a `make format` and a fresh rebase as well. Passes the full `driver.pl` on my local Ubuntu in docker container: 
`TESTS DONE, PASSED: Passed 2024  Failed 0  Time 123:45`
(By the way just an FYI, I natively run Fedora37, for which I was able to get all Verilator packages installed and working correctly, except for one: clang-format-14. I believe because F37 has moved to clang 15.)